### PR TITLE
Reduce memory consumption for string concatenation from O(n^2) to O(n)

### LINKED ISSUE #7398: OOM during constant string folding due to allocations from ConstantStringValue
Currently in ASP.NET Razor code generation, when we have a large contiguous block of HTML, it gets code-generated as a series of concatenated string constants:

Example:

```
        #pragma warning disable 1998
        public override async Task ExecuteAsync()
        {
            BeginContext(48, 31408, true);
            WriteLiteral(
"<!DOCTYPE html><html prefix=\"og: http://ogp.me/ns# fb: http://ogp.me/ns/fb#\"  lan" +
"g=\"en-US\"  style=\"font-size:62.5%\"  class=\"hiperf\"  dir=\"ltr\" >\r\n\r\n<head data-in" +
"fo=\"v:2.0.5798.31020;a:cc7227b7-f7fe-4e3b-b644-226f3ef2c489;cn:125;az:{did:1f802" +
"c514e1c42e385b028b15ac20ce4, rid: 125, sn: westus-hp, dt: 2015-11-16T22:01:08.08" +
"25616Z, bt: 2015-11-17T01:14:47.5617874Z};ddpi:1;dpio:;dpi:1;dg:tmx.pc.webkit.ch" +
"rome;th:start;PageName:startPage;m:en-us;cb:;l:en-us;mu:en-us;ud:{cid:,vk:homepa" +
"ge,n:,l:en-us,ck:};xd:AA3iOMG;ovc:f;al:;fxd:t;xdpub:2015-11-18 20:26:50Z;xdmap:2" +
"015-11-18 20:30:01Z;axd:;f:muidflt2cf,muidflt9-ismsft,d-thshldie1,d-thshldie2\"  " +
"data-js=\"{&quot;dpi&quot;:1.0,&quot;ddpi&quot;:1.0,&quot;dpio&quot;:null,&quot;f" +
"orcedpi&quot;:null,&quot;dms&quot;:6000,&quot;bds&quot;:7,&quot;dg&quot;:&quot;t" +
"mx.pc.webkit.chrome&quot;,&quot;ssl&quot;:false,&quot;moduleapi&quot;:&quot;http" +
"://www.msn.com/en-us/homepage/api/modules/fetch&quot;,&quot;cdnmoduleapi&quot;:&" +
"quot;http://static-hp-wus.s-msn.com/en-us/homepage/api/modules/cdnfetch&quot;,&q" +
"uot;pdpdeltaupdateapi&quot;:&quot;http://www.msn.com/en-us/homepage/api/pdp/upda" +
"tepdpdata&quot;,&quot;xd&quot;:&quot;aa3iomg&quot;,&quot;signedin&quot;:1,&quot;" +
"sso&quot;:&quot;&quot;,&quot;exchangeenabled&quot;:true,&quot;twitterimpenabled&" +
...
```

For example, taking the home page of MSN.com and running it through Razor yields something like https://gist.github.com/rynowak/de862bfb9199caa578f8

This causes some really intense allocation during Roslyn compilation of the generated code, and will OOM. 

image
# 

On our side we're going to take some steps to work around this - the way the codegen works today is largely a product of legacy carried forward from CodeDOM :laughing: In the end, we'll still likely be dealing with a need to efficiently codegen C# and compile it using Roslyn during **runtime**, so any advice you have would be beneficial.

Even though the `string.Concat`s happen at compile-time we still end up with an LOH string at runtime, which we'd like to avoid.

We're going to start by getting rid of the `string.Concat`s here, and just emit strings of a non-LOH size with more calls to `WriteLiteral`.

### LINKED ISSUE #37572: Error with Large HTML
**Version Used**: 
VS2019 COMMUNITY

**Steps to Reproduce**:
1. FIRST CREATE A VB WEB PROJECT WEB FORMS
2. CREATE PAGE1.ASPX
3. WRITE A LARGE HTML IN IT
4. TRY TO RUN THE PROJECT

**Expected Behavior**:
SEE HTML NORMALLY

**Actual Behavior**:
ERROR THAT APPEAR:
https://github.com/rogeriorlima3/errorVS2019/issues/1

### PR DESCRIPTION
**Pull request by gafter**
Fixes #7398
Fixes #37572
Relates to https://github.com/aspnet/Razor/issues/614 and many other customer accomodations that are no longer needed.

History of this bug:  When constant folding a long sequence of string concatentations, there is an intermediate constant value for every left-hand operand.  So the total memory consumed to compute the whole concatenation was *O(n^2)*.  The compiler would simply perform this work and eventually run out of memory, simply crashing with no useful diagnostic.  Later, the concatenation implementation was instrumented so it would detect when it was likely to run out of memory soon, and would instead report a diagnostic at the last step. See https://github.com/dotnet/roslyn/commit/f177077665a5bef862014f1898b961d1759c9248.  Test `Bug529600()` was added to demonstrate that we produced a diagnostic.  However, the compiler still consumed *O(n^2)* memory for the concatenation and this test used to consume so much memory that it would cause other tests running in parallel to fail because they might not have enough memory to succeed.  So the test was disabled and eventually removed.  The compiler would still crash with programs containing large string concatenations, or consume huge amounts of memory and take a long time before reporting a diagnostic, so the underlying problem had not been addressed.

Here we have revised the implementation of constant folding string concatenations so that it requires *O(n)* memory and remove the old instrumentation.  As a consequence the test `Bug529600()` now runs very quickly and does not consume gobs of memory.

### CODE REVIEW

File: src/Compilers/Core/Portable/Collections/Rope.cs [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+                return false;
+            if (Length == 0)
+                return true;
+            var chars0 = GetChars().GetEnumerator();
+            var chars1 = other.GetChars().GetEnumerator();
+            while (chars0.MoveNext() && chars1.MoveNext())
+            {
+                if (chars0.Current != chars1.Current)
+                    return false;
+            }
```
Comment (rynowak) [inline src/Compilers/Core/Portable/Collections/Rope.cs]: won't this loop bail early if one of the ropes is longer than the other? 'AAA' vs 'AA'

Comment (RikkiGibson) [reply src/Compilers/Core/Portable/Collections/Rope.cs]: seems like we already ensured the two ropes have equal lengths by this point.

**Comment by 333fred** [comment]
> // Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.

Consider enabling nullable for this file, since it's a totally new class.

---
Refers to: src/Compilers/Core/Portable/Collections/Rope.cs:1 in a750642. [](commit_id = a75064262bc35d4c58a9c77c30253888853b0260, deletion_comment = False)

File: src/Compilers/Core/Portable/Collections/Rope.cs [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+
+            return true;
+        }
+        public override int GetHashCode()
+        {
+            int result = Length;
+            foreach (char c in GetChars())
+                result = Hash.Combine((int)c, result);
```
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/Collections/Rope.cs]: note: there are ways you can design ropes such that the sub parts contain their hashes, and they can be combined such that any equivalent ropes would have the same hash (i did this in my Rope implementation at Google :)).

If this is an operation called a lot, that may be worth doing.  However, if the compiler has no need to hash these, then that work isn't necessary.

File: src/Compilers/Core/Portable/Collections/Rope.cs [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+            var chars1 = other.GetChars().GetEnumerator();
+            while (chars0.MoveNext() && chars1.MoveNext())
+            {
+                if (chars0.Current != chars1.Current)
+                    return false;
+            }
+
+            return true;
+        }
+        public override int GetHashCode()
```
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/Collections/Rope.cs]:
```suggestion

        public override int GetHashCode()
```

File: src/Compilers/Core/Portable/Collections/Rope.cs [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+                            throw ExceptionUtilities.UnexpectedValue(v.GetType().Name);
+                    }
+                }
+
+                return psb.ToStringAndFree();
+            }
+
+            protected override IEnumerable<char> GetChars()
+            {
+                var stack = new Stack<Rope>();
```
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/Collections/Rope.cs]: if GetChars is called many times, consider pooling this stack.

File: src/Compilers/Core/Portable/ConstantValueSpecialized.cs [hunk 5]
```diff
@@ -85,6 +85,14 @@ public override string StringValue
                 }
             }
 
+            internal override Rope RopeValue
```
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/ConstantValueSpecialized.cs]:
```suggestion
            internal override Rope RopeValue => null;
```

File: src/Compilers/Core/Portable/Collections/Rope.cs [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+{
+    /// <summary>
+    /// A representation of a string of characters that requires O(1) extra space to concatenate two ropes.
+    /// </summary>
+    internal abstract class Rope
+    {
+        public static readonly Rope Empty = ForString("");
+        public abstract override string ToString();
+        public abstract int Length { get; }
+        protected abstract IEnumerable<char> GetChars();
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/Collections/Rope.cs]: Did you consider using a `Span<char>` here given this type is meant to help with perf scenarios?

File: src/Compilers/Core/Portable/Collections/Rope.cs [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+        /// Two ropes are "the same" if they represent the same sequence of characters.
+        /// </summary>
+        public override bool Equals(object obj)
+        {
+            if (!(obj is Rope other) || Length != other.Length)
+                return false;
+            if (Length == 0)
+                return true;
+            var chars0 = GetChars().GetEnumerator();
+            var chars1 = other.GetChars().GetEnumerator();
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/Collections/Rope.cs]: Implementing `GetChars` in terms of `Span<char>` would avoid allocations in this method and let you use `Span<T>.SequenceEquals` which is very efficient on CoreClr

File: src/Compilers/Core/Portable/Collections/Rope.cs [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+            foreach (char c in GetChars())
+                result = Hash.Combine((int)c, result);
+
+            return result;
+        }
+
+        /// <summary>
+        /// A rope that wraps a simple string.
+        /// </summary>
+        private class StringRope : Rope
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/Collections/Rope.cs]: `sealed`

File: src/Compilers/Core/Portable/Collections/Rope.cs [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+            public StringRope(string value) => _value = value;
+            public override string ToString() => _value;
+            public override int Length => _value.Length;
+            protected override IEnumerable<char> GetChars() => _value;
+        }
+
+        /// <summary>
+        /// A rope that represents the concatenation of two ropes.
+        /// </summary>
+        private class ConcatRope : Rope
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/Collections/Rope.cs]: `sealed`

File: src/Compilers/Core/Portable/Collections/Rope.cs [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+                            break;
+                        case var v:
+                            throw ExceptionUtilities.UnexpectedValue(v.GetType().Name);
+                    }
+                }
+
+                return psb.ToStringAndFree();
+            }
+
+            protected override IEnumerable<char> GetChars()
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/Collections/Rope.cs]: Hmm, this method is going to make it hard to have a `Span<char>` method here :frown:

File: src/Compilers/Core/Portable/Collections/Rope.cs (hunk 2 above)
Comment (gafter) [reply src/Compilers/Core/Portable/Collections/Rope.cs]: The compiler hashes these, but rarely.  The only times it does is when a string appears as a case in a switch.

---
In reply to: [313170138](https://github.com/dotnet/roslyn/pull/37915#discussion_r313170138) [](ancestors = 313170138)

File: src/Compilers/Core/Portable/Collections/Rope.cs (hunk 4 above)
Comment (gafter) [reply src/Compilers/Core/Portable/Collections/Rope.cs]: It isn't.

---
In reply to: [313170423](https://github.com/dotnet/roslyn/pull/37915#discussion_r313170423) [](ancestors = 313170423)

File: src/Compilers/Core/Portable/ConstantValueSpecialized.cs (hunk 5 above)
Comment (gafter) [reply src/Compilers/Core/Portable/ConstantValueSpecialized.cs]: I would do that if all the surrounding ones used that style.

---
In reply to: [313170693](https://github.com/dotnet/roslyn/pull/37915#discussion_r313170693) [](ancestors = 313170693)

File: src/Compilers/Core/Portable/Collections/Rope.cs (hunk 2 above)
Comment (gafter) [reply src/Compilers/Core/Portable/Collections/Rope.cs]: Removing most of those now.

---
In reply to: [313902881](https://github.com/dotnet/roslyn/pull/37915#discussion_r313902881) [](ancestors = 313902881,313170138)

**Comment by dotcom9** [comment]
Still getting this error trying to build an ASP.NET Core 3.1 application in VS2019 16.4.457.38025.

**Error CS8103 Combined length of user strings used by the program exceeds allowed limit. Try to decrease use of string literals.**

**Comment by gafter** [comment]
@dotcom9 You are probably seeing that error not because concatenation is implemented inefficiently, but because your program genuinely has more string literals than can be represented in the generated assembly.

**Comment by dotcom9** [comment]
> @dotcom9 You are probably seeing that error not because concatenation is implemented inefficiently, but because your program genuinely has more string literals than can be represented in the generated assembly.

It doesn't appear to have anything to do with string literals if you're referring to literals hard-coded in classes. It has to do with views. If I remove a folder of views from the application, it compiles perfectly. The use of string literals in the views is minimal. I assume the compiler does not include content strings / text as string literals?

**Comment by gafter** [comment]
C# does not have a concept of "views".  I suspect that ASP.net does indeed compile them into string literals.

**Comment by robinwilson16** [comment]
I can confirm this is not fixed and still occurs today in .NET 7. Removing a folder of razor pages causes the project to compile and it doesn't matter which pages are removed, there are just too many for it to handle but if you have a large website and want to keep it all then I'm not sure what can be done except to split the site into smaller sub-sites but that is far from ideal.

**Comment by CyrusNajmabadi** [comment]
@robinwilson16 can you move from large string literals to resources instead?

**Comment by 333fred** [comment]
@robinwilson16 can you clarify what you mean by the project not successfully compiling?

**Comment by AlekseyTs** [comment]
@robinwilson16 

> I can confirm this is not fixed and still occurs today in .NET 7.

Are you referring to the issue reported by https://github.com/dotnet/razor/issues/7842? This PR was never meant to deal with that error and is about a different problem.

**Comment by robinwilson16** [comment]
> @robinwilson16 can you move from large string literals to resources instead?

I was converting an old website with 354 static pages to .NET although I do plan to move key pages to being database driven but it would be a big job to move everything again manually. Runtime compilation was a workaround and this is working acceptably quickly for now.

**Comment by robinwilson16** [comment]
> @robinwilson16
> 
> > I can confirm this is not fixed and still occurs today in .NET 7.
> 
> Are you referring to the issue reported by [dotnet/razor#7842](https://github.com/dotnet/razor/issues/7842)? This PR was never meant to deal with that error and is about a different problem.

Yes sorry I now realise these appear different and it was in fact the other issue I was experiencing which it appear has not been addressed as is only experienced by a small number of people as most do not have so many pages. All pages contain either a lot of text or large tables of data and the limit was hit around the 50% point so I had to stop and find a workaround to be able to continue previewing the site as I built it.
Sorry for any confusion.

**Comment by CyrusNajmabadi** [comment]
>but it would be a big job to move everything again manually.

I would recommend writing a tool to do it.  So that you don't have to do each string manually :)

**Comment by robinwilson16** [comment]
It would currently involve a resource for each page though so would double the number of files and as older pages do not need amending and new pages will be moved to the database and runtime compilation works ok then I think I may just leave it as is unless there would be a benefit to doing this.
Hopefully the issue will be fixed for .NET 8 and I can revert to pre-compiled pages to reduce my IOPS/cost a little.

**Comment by huzhuozhuo** [comment]
This problem has not been solved yet. What method can be used to run it

**Pull request merged**